### DISCUSSION TITLE
Best way to implement reusable Custom Field Types?

### DISCUSSION THREAD

**Discussion by deric4** [aws-powertools/powertools-lambda-python#555] [Q&A] [answered]
I'm attempting  to add some [Custom Field Types](https://pydantic-docs.helpmanual.io/usage/types/#custom-data-types) like `arn`, `instance-id`, `snapshot-id`, `account-number`, `ami-id`, across multiple projects (similar to [Pydantic Types](https://pydantic-docs.helpmanual.io/usage/types/#pydantic-types) like `EmailStr` or `NameEmail`)
 
Still in the early stages of this, but was planning on implementing these field types as a standalone library so they can be used in projects that don't use **aws-lambda-powertools-python** (i.e. fastapi) and came across this in the [Parser FAQ section](https://awslabs.github.io/aws-lambda-powertools-python/latest/utilities/parser/#faq) of the docs:

<img width="677" alt="escape-hatch" src="https://user-images.githubusercontent.com/5762138/126729088-6191cb5f-fa80-4f24-9655-d1bf16d9fe1c.png">

Based on the above, is using things like Pydantic's builtin classes like `errors.StrRegexError` possible?
<img width="648" alt="pydantic-errors" src="https://user-images.githubusercontent.com/5762138/126730171-68d2100a-8fd3-4dc6-ad82-727231f5cce3.png">
https://github.com/samuelcolvin/pydantic/blob/master/pydantic/types.py#L401-L409

<br>
<br>

If so, would I need to define the import statement need to be different depending on if the library was being used in a powertools lambda function or not?

Was also looking at [Constrained Types](https://pydantic-docs.helpmanual.io/usage/types/#constrained-types) like [Netflix ConsoleMe](https://github.com/Netflix/consoleme/blob/master/consoleme/models.py#L27-L29) but still not sure if there would be probs with the `import` statements.

<br>
<br>

*somewhat related Pydantic GH Issue:*
https://github.com/samuelcolvin/pydantic/issues/1593

**Comment by heitorlessa** [comment, marked as answer]
hey @deric4 somehow we missed this due to GitHub Discussions feature. In your case, it's best to use `import pydantic`, because of this line `so they can be used in projects that don't use aws-lambda-powertools-python`.

What we're doing in Lambda Powertools is making sure you have an escape hatch should you need something from `pydantic` that we don't expose directly. Because you have to install `pydantic` to use the Parser utility, this will be available in the import namespace, so you can use `from pydantic import ....`, or `import pydantic` as you normally would.

**Reply by deric4** [reply]
awesome. thanks @heitorlessa
